plugins/procedural/asset_utils.cpp

```diff
@@ -0,0 +1,509 @@
+    std::vector<USDDependency>& dependencies, UsdStageRefPtr stage, 
+    const SdfLayerHandle& layer, ArResolver& resolver, 
+    std::unordered_set<std::string>& seenReferences)
+{
+    if (ref.empty())
+        return;
+
+    if (seenReferences.find(ref) != seenReferences.end())
+        return;
+
+    // resolve the reference to an absolute path
+    std::string anchoredPath = SdfComputeAssetPathRelativeToLayer(layer, ref);
+    std::string resolvedPath = resolver.Resolve(anchoredPath);
+    // convert a relative reference relative to the main scene
+    std::string finalRefPath = ref;
+    if (!resolvedPath.empty() && std::filesystem::path(ref).is_relative())
+    {
+        std::string relativeToRoot = ComputeRelativePathToRoot(stage, resolvedPath);
+        // convert only if the file is located under the root folder
+        if (!relativeToRoot.empty() && relativeToRoot.at(0) != '.')
+            finalRefPath = relativeToRoot;
+    }
+
+    // create a dependency
+    dependencies.push_back(USDDependency(depType, finalRefPath, 
+        resolvedPath, layer->GetIdentifier(), primPath, attribute));
+    seenReferences.insert(ref);
+}
+
+/**
+ * Helper function to iterate over all prims in a layer.
+ */
+inline void TraversePrimSpecs(
+    const SdfPrimSpecHandle& prim,
+    const std::function<void(const SdfPrimSpecHandle&)>& fn)
+{
+    fn(prim);
+
+    for (const auto& child : prim->GetNameChildren())
+        TraversePrimSpecs(child, fn);
+}
+
+/**
+ * Helper function to iterate over all prims in a layer.
+ */
+inline void TraverseLayer(const SdfLayerHandle& layer,
+                   const std::function<void(const SdfPrimSpecHandle&)>& fn)
+{
+    SdfPrimSpecHandle root = layer->GetPseudoRoot();
+
+    for (const auto& prim : root->GetNameChildren())
+        TraversePrimSpecs(prim, fn);
+}
+
+/**
+ * Returns all dependencies found in the given layer.
+ */
+inline void CollectDependenciesFromLayer(
+    UsdStageRefPtr stage,
+    const SdfLayerHandle& layer,
+    std::vector<USDDependency>& dependencies,
+    std::unordered_set<std::string>& seenReferences,
+    ArResolver& resolver)
+{
+    if (!layer)
+        return;
+
+    // collect sublayers
+    for (const std::string& sub : layer->GetSubLayerPaths())
+    {
+        AddDependency(sub, "sublayer", layer->GetDisplayName(),
+            "sublayer", dependencies, stage, layer, resolver, seenReferences);
+    }
+
+    // iterate all prims in this layer
+    TraverseLayer(layer, [&](const SdfPrimSpecHandle& primSpec) {
+        const std::string primPath = primSpec->GetPath().GetString();
+
+        // collect dependencies from attributes
+        for (const SdfAttributeSpecHandle& attr : primSpec->GetAttributes())
+        {
+            if (!attr)
+                continue;
+
+            const std::string attrName = attr->GetName();
+
+            // asset type attribute
+            if (attr->GetTypeName() == SdfValueTypeNames->Asset)
```

Comment:
We might have to test the strings as well, as some of arnold assets are stored as string, let me find which one and get back to you